File: ondo-cli/src/library.rs
//! Il menu Libreria: le viste, e le azioni su un singolo video.

use console::style;
use dialoguer::{Confirm, Select};
use ondo::{Filter, Mode, State};

use crate::ui::{self, screen, R};
use crate::App;

pub fn open(app: &mut App) -> R<()> {
    loop {
        app.pump(|_| {});
        screen("Libreria", &mut app.message);

        let da_scaricare = app.lib.count(Filter::Pending) + app.lib.queue().iter().filter(|q| q.error.is_none()).count();
        let falliti = app.lib.count(Filter::Failed) + app.lib.queue().iter().filter(|q| q.error.is_some()).count();
        let voci = vec![
            format!("Tutti i video ({})", app.lib.count(Filter::All)),
            format!("Autori ({})", app.lib.authors().len()),
            format!("Preferiti ({})", app.lib.count(Filter::Favorites)),
            format!("Da scaricare ({da_scaricare})"),
            format!("Archiviati ({})", app.lib.count(Filter::Archived)),
            format!("Rimossi da YouTube ({})", app.lib.count(Filter::Removed)),
            format!("Falliti ({falliti})"),
            "← indietro".to_string(),
        ];

        match Select::with_theme(&ui::theme()).items(&voci).default(0).interact_opt()? {
            Some(0) => list_view(app, Filter::All)?,
            Some(1) => authors_view(app)?,
            Some(2) => list_view(app, Filter::Favorites)?,
            Some(3) => pending_view(app)?,
            Some(4) => list_view(app, Filter::Archived)?,
            Some(5) => list_view(app, Filter::Removed)?,
            Some(6) => failed_view(app)?,
            _ => return Ok(()),
        }
    }
}

/// Una vista semplice: elenco di video, si scelgono per agire.
fn list_view(app: &mut App, filter: Filter) -> R<()> {
    loop {
        app.pump(|_| {});
        let titolo = format!("Libreria · {}", filter.label());
        screen(&titolo, &mut app.message);

        let video = app.lib.list(filter);
        if video.is_empty() {
            println!("Non c'è niente qui.\n");
        }
        let ids: Vec<String> = video.iter().map(|v| v.id.clone()).collect();
        let mut voci: Vec<String> = video.iter().map(|v| ui::video_line(v)).collect();
        voci.push("← indietro".into());

        match Select::with_theme(&ui::theme()).items(&voci).default(0).max_length(15).interact_opt()? {
            Some(i) if i < ids.len() => actions(app, &ids[i])?,
            _ => return Ok(()),
        }
    }
}

fn authors_view(app: &mut App) -> R<()> {
    loop {
        app.pump(|_| {});
        screen("Libreria · autori", &mut app.message);

        let autori = app.lib.authors();
        let mut voci: Vec<String> = autori.iter().map(|(a, n)| format!("{a}  ({n})")).collect();
        voci.push("← indietro".into());

        let scelta = Select::with_theme(&ui::theme()).items(&voci).default(0).max_length(15).interact_opt()?;
        let Some(i) = scelta.filter(|i| *i < autori.len()) else { return Ok(()) };
        let autore = autori[i].0.clone();

        loop {
            app.pump(|_| {});
            screen(&format!("Libreria · {autore}"), &mut app.message);
            let video = app.lib.by_author(&autore);
            let ids: Vec<String> = video.iter().map(|v| v.id.clone()).collect();
            let mut voci: Vec<String> = video.iter().map(|v| ui::video_line(v)).collect();
            voci.push("← indietro".into());

            match Select::with_theme(&ui::theme()).items(&voci).default(0).max_length(15).interact_opt()? {
                Some(i) if i < ids.len() => actions(app, &ids[i])?,
                _ => break,
            }
        }
    }
}

/// "Da scaricare" mette insieme due cose che per l'utente sono la stessa: i video
/// risolti che aspettano il disco, e i link appena incollati che non hanno ancora
/// un id.
fn pending_view(app: &mut App) -> R<()> {
    loop {
        app.pump(|_| {});
        screen("Libreria · da scaricare", &mut app.message);

        // Chi sta scaricando adesso va in cima e lo dice: sta succedendo, non è in
        // attesa come gli altri. `sort_by_key` è stabile, quindi dentro i due
        // gruppi resta l'ordine di `list`.
        let mut video = app.lib.list(Filter::Pending);
        video.sort_by_key(|v| v.state != State::Downloading);

        let ids: Vec<String> = video.iter().map(|v| v.id.clone()).collect();
        let mut voci: Vec<String> = video
            .iter()
            .map(|v| {
                let tag = if v.state == State::Downloading { "In download" } else { "In attesa" };
                ui::video_line_tagged(v, tag)
            })
            .collect();
        let links: Vec<String> = app
            .lib
            .queue()
            .iter()
            .filter(|q| q.error.is_none())
            .map(|q| q.url.clone())
            .collect();
        for url in &links {
            voci.push(format!("{} {url}  {}", style("○").yellow(), style("da risolvere").cyan()));
        }

        let scarica_tutti = !ids.is_empty() || !links.is_empty();
        if scarica_tutti {
            voci.insert(0, format!("Scarica tutti ({})", ids.len() + links.len()));
        } else {
            println!("Niente in attesa.\n");
        }
        voci.push("← indietro".into());

        let scelta = Select::with_theme(&ui::theme()).items(&voci).default(0).max_length(15).interact_opt()?;
        let offset = usize::from(scarica_tutti);
        match scelta {
            Some(0) if scarica_tutti => {
                // La risoluzione si chiede **una volta** per tutto il gruppo:
                // chiederla per ognuno sarebbe un interrogatorio.
                let Some(cap) = ui::resolve_quality(app.lib.config().quality, "Che risoluzione, per tutti?")?
                else {
                    continue;
                };
                let mut n = 0;
                let mut in_corso = 0;
                for id in &ids {
                    // Chi sta già scaricando si lascia in pace: rimetterlo in coda
                    // vorrebbe dire un secondo sentinel sullo stesso video.
                    if app.lib.get(id).map(|v| v.state == State::Downloading).unwrap_or(false) {
                        in_corso += 1;
                        continue;
                    }
                    if let Ok(Some(url)) = app.lib.retry(id) {
                        if !url.is_empty() {
                            app.dl.push_with(url, cap);
                            n += 1;
                        }
                    }
                }
                for url in &links {
                    app.dl.push_with(url.clone(), cap);
                    n += 1;
                }
                let salta = if in_corso > 0 { format!(" ({in_corso} già in corso, lasciati stare)") } else { String::new() };
                ui::ok(&mut app.message, format!("{n} accodati{salta} — guardali in «Download rapido»"));
            }
            Some(i) if i >= offset && i - offset < ids.len() => actions(app, &ids[i - offset].clone())?,
            Some(i) if i >= offset + ids.len() && i - offset - ids.len() < links.len() => {
                let url = links[i - offset - ids.len()].clone();
                link_actions(app, &url)?;
            }
            _ => return Ok(()),
        }
    }
}

/// I falliti, con il loro motivo: video che hanno provato e link che non si sono
/// nemmeno lasciati risolvere.
fn failed_view(app: &mut App) -> R<()> {
    loop {
        app.pump(|_| {});
        screen("Libreria · falliti", &mut app.message);

        let video = app.lib.list(Filter::Failed);
        let ids: Vec<String> = video.iter().map(|v| v.id.clone()).collect();
        // Una voce di menu resta su **una** riga: `dialoguer` misura le voci per
        // disegnare la selezione, e una voce a due righe gli fa perdere il conto.
        let mut voci: Vec<String> = video
            .iter()
            .map(|v| {
                format!(
                    "{}  {}",
                    ui::video_line(v),
                    style(motivo(v.error.as_deref())).red().dim()
                )
            })
            .collect();
        let links: Vec<String> = app
            .lib
            .queue()
            .iter()
            .filter(|q| q.error.is_some())
            .map(|q| q.url.clone())
            .collect();
        for q in app.lib.queue().iter().filter(|q| q.error.is_some()) {
            voci.push(format!(
                "{} {}  {}",
                style("✗").red(),
                q.url,
                style(motivo(q.error.as_deref())).red().dim()
            ));
        }
        if voci.is_empty() {
            println!("Nessun fallimento.\n");
        }
        voci.push("← indietro".into());

        match Select::with_theme(&ui::theme()).items(&voci).default(0).max_length(15).interact_opt()? {
            Some(i) if i < ids.len() => actions(app, &ids[i].clone())?,
            Some(i) if i - ids.len() < links.len() => {
                let url = links[i - ids.len()].clone();
                link_actions(app, &url)?;
            }
            _ => return Ok(()),
        }
    }
}

/// Il motivo di un fallimento ridotto a una riga: quello che serve per capire
/// *quale* fallimento è. Per intero si vede in «Dettagli».
fn motivo(error: Option<&str>) -> String {
    let testo = error.unwrap_or("motivo non registrato").replace(['\n', '\r'], " ");
    // Il pezzo utile di un errore di yt-dlp sta dopo `ERROR:`.
    let testo = match testo.split_once("ERROR:") {
        Some((_, dopo)) => dopo.trim().to_string(),
        None => testo,
    };
    if testo.chars().count() > 60 {
        format!("{}…", testo.chars().take(59).collect::<String>())
    } else {
        testo
    }
}

/// Cosa si può fare con un link che non è (ancora) un video.
fn link_actions(app: &mut App, url: &str) -> R<()> {
    screen("Link in coda", &mut app.message);
    println!("{url}\n");
    let voci = ["Riprova adesso", "Toglilo dalla coda", "← indietro"];
    match Select::with_theme(&ui::theme()).items(&voci).default(0).interact_opt()? {
        Some(0) => {
            let Some(cap) = ui::resolve_quality(app.lib.config().quality, "Che risoluzione?")? else {
                return Ok(());
            };
            app.dl.push_with(url.to_string(), cap);
            ui::ok(&mut app.message, "accodato");
        }
        Some(1) => {
            app.lib.dequeue(url)?;
            ui::ok(&mut app.message, "togliato dalla coda");
        }
        _ => {}
    }
    Ok(())
}

/// Le azioni su un video, contestuali al suo stato. Usata anche dalla ricerca:
/// un solo posto dove si decide cosa si può fare con un video.
///
/// Le statistiche stanno **in alto**, sopra le voci: è quello che si vuole sapere
/// prima di decidere, e rende inutile una voce «Dettagli».
pub fn actions(app: &mut App, id: &str) -> R<()> {
    loop {
        app.pump(|_| {});
        let Some(v) = app.lib.get(id).cloned() else {
            return Ok(());
        };
        screen(&format!("{} — {}", v.author, v.title), &mut app.message);
        statistiche(app, &v);

        // Le voci si costruiscono in base allo stato: niente azioni impossibili da
        // scegliere e poi rifiutare. Senza icone, come il resto dei menu.
        let mut voci: Vec<String> = Vec::new();
        let mut azioni: Vec<Azione> = Vec::new();
        if v.state == State::Downloaded {
            voci.push("Riproduci".into());
            azioni.push(Azione::Play);
            voci.push("Solo audio".into());
            azioni.push(Azione::Audio);
        }
        if matches!(v.state, State::Pending | State::Failed) {
            voci.push("Scarica ora".into());
            azioni.push(Azione::Scarica);
        }
        voci.push(if v.favorite { "Togli dai preferiti".into() } else { "Aggiungi ai preferiti".to_string() });
        azioni.push(Azione::Preferito(!v.favorite));
        voci.push(if v.archived { "Ripristina".into() } else { "Archivia".to_string() });
        azioni.push(Azione::Archivia(!v.archived));
        voci.push("Togli dalla libreria".into());
        azioni.push(Azione::Rimuovi);
        voci.push("← indietro".into());

        let scelta = Select::with_theme(&ui::theme()).items(&voci).default(0).interact_opt()?;
        let Some(i) = scelta.filter(|i| *i < azioni.len()) else { return Ok(()) };

        match azioni[i] {
            Azione::Play => match app.lib.play(id, Mode::Video) {
                Ok(()) => ui::ok(&mut app.message, "VLC avviato"),
                Err(e) => ui::err(&mut app.message, e),
            },
            Azione::Audio => match app.lib.play(id, Mode::Audio) {
                Ok(()) => ui::ok(&mut app.message, "VLC avviato (solo audio)"),
                Err(e) => ui::err(&mut app.message, e),
            },
            Azione::Scarica => {
                // Se la qualità è «chiedi ogni volta», si chiede adesso: è il
                // momento in cui il download parte.
                let Some(cap) = ui::resolve_quality(app.lib.config().quality, "Che risoluzione?")? else {
                    continue;
                };
                match app.lib.retry(id)? {
                    Some(url) if !url.is_empty() => {
                        app.dl.push_with(url, cap);
                        ui::ok(&mut app.message, "accodato — guardalo in «Download rapido»");
                    }
                    _ => ui::err(&mut app.message, "questo video non ha un URL da riscaricare"),
                }
            }
            Azione::Preferito(on) => {
                app.lib.set_favorite(id, on)?;
                ui::ok(&mut app.message, if on { "nei preferiti" } else { "togliato dai preferiti" });
            }
            Azione::Archivia(on) => {
                app.lib.set_archived(id, on)?;
                ui::ok(&mut app.message, if on { "archiviato" } else { "ripristinato" });
            }
            Azione::Rimuovi => {
                let conferma = Confirm::with_theme(&ui::theme())
                    .with_prompt(format!("Togliere «{}» dalla libreria?", v.title))
                    .default(false)
                    .interact()?;
                if conferma {
                    let anche_file = v.state == State::Downloaded
                        && Confirm::with_theme(&ui::theme())
                            .with_prompt("Cancellare anche il file dal disco?")
                            .default(false)
                            .interact()?;
                    app.lib.remove(id, anche_file)?;
                    ui::ok(
                        &mut app.message,
                        if anche_file { "togliato, file cancellato" } else { "togliato dalla libreria" },
                    );
                    return Ok(());
                }
            }
        }
    }
}

#[derive(Clone, Copy)]
enum Azione {
    Play,
    Audio,
    Scarica,
    Preferito(bool),
    Archivia(bool),
    Rimuovi,
}

/// Le statistiche del video, sopra le voci del menu.
fn statistiche(app: &App, v: &ondo::Video) {
    let riga = |k: &str, val: String| println!("  {:<12} {val}", style(k).dim());
    riga("stato", style(v.state.label()).cyan().to_string());
    riga("durata", v.duration_label());
    riga("pubblicato", v.upload_date.clone().unwrap_or_else(|| "—".into()));
    riga(
        "risoluzione",
        match (v.width, v.height) {
            (Some(w), Some(h)) => {
                let fps = v.fps.map(|f| format!(" @ {f:.0}fps")).unwrap_or_default();
                format!("{w}×{h}{fps}")
            }
            _ => "—".into(),
        },
    );
    if v.size_bytes > 0 {
        riga("dimensione", ui::size_label(v.size_bytes));
    }
    riga("url", v.url.clone());
    if v.state == State::Downloaded {
        let file = app.lib.file_path(v);
        let manca = if file.is_file() { String::new() } else { style("  (non è al suo posto)").red().to_string() };
        riga("file", format!("{}{manca}", file.display()));
    }
    if !v.tags.is_empty() {
        let tag: Vec<&str> = v.tags.iter().take(8).map(String::as_str).collect();
        riga("tag", tag.join(", "));
    }
    if let Some(e) = &v.error {
        riga("errore", style(e).red().to_string());
        riga("tentativi", v.attempts.to_string());
    }
    println!();
}

File: packages/ondo-cli/src/menu/library.js
// M86 — traduzione di `cli/src/library.rs`.
//
// Il menu Libreria: le viste, e le azioni su un singolo video.

import { existsSync } from 'node:fs';

import * as core from '@catalog/core';
import * as ui from './ui.js';
import { FILTER, STATE } from './ondo.js';

const BACK = Symbol('back');

// `confirm` che si può annullare come nell'originale: Esc/Ctrl-C valgono "no"
// (M88: l'Esc passa da `ui.confirmOpt`, che ritorna `null` per "annullato" —
// qui, su una domanda sì/no, annullare *è* rispondere no).
async function confirmOpt(message, defaultValue = false) {
  return (await ui.confirmOpt({ message, defaultValue })) === true;
}

export async function open(app) {
  while (true) {
    await app.pump();
    ui.screen('Libreria', app);

    const daScaricare =
      app.lib.count(FILTER.PENDING) + app.lib.queue().filter((q) => !q.error).length;
    const falliti = app.lib.count(FILTER.FAILED) + app.lib.queue().filter((q) => q.error).length;

    const voci = [
      { name: `Tutti i video (${app.lib.count(FILTER.ALL)})`, value: FILTER.ALL },
      { name: `Autori (${app.lib.authors().length})`, value: 'authors' },
      { name: `Preferiti (${app.lib.count(FILTER.FAVORITES)})`, value: FILTER.FAVORITES },
      { name: `Da scaricare (${daScaricare})`, value: 'pending' },
      { name: `Archiviati (${app.lib.count(FILTER.ARCHIVED)})`, value: FILTER.ARCHIVED },
      { name: `Rimossi da YouTube (${app.lib.count(FILTER.REMOVED)})`, value: FILTER.REMOVED },
      { name: `Falliti (${falliti})`, value: 'failed' },
      { name: '← indietro', value: BACK }
    ];

    const scelta = await ui.selectOpt({ choices: voci, defaultValue: voci[0].value, pageSize: voci.length });
    if (scelta === null || scelta === BACK) return;
    if (scelta === 'authors') await authorsView(app);
    else if (scelta === 'pending') await pendingView(app);
    else if (scelta === 'failed') await failedView(app);
    else await listView(app, scelta);
  }
}

/** Una vista semplice: elenco di video, si scelgono per agire. */
async function listView(app, filter) {
  while (true) {
    await app.pump();
    ui.screen(`Libreria · ${core.filterLabel(filter)}`, app);

    const video = app.lib.list(filter);
    if (video.length === 0) console.log("Non c'è niente qui.\n");

    const voci = [
      ...video.map((v) => ({ name: ui.videoLine(v), value: v.id })),
      { name: '← indietro', value: BACK }
    ];

    const scelta = await ui.selectOpt({ choices: voci, defaultValue: voci[0].value });
    if (scelta === null || scelta === BACK) return;
    await actions(app, scelta);
  }
}

async function authorsView(app) {
  while (true) {
    await app.pump();
    ui.screen('Libreria · autori', app);

    const autori = app.lib.authors();
    const voci = [
      ...autori.map((a) => ({ name: `${a.author}  (${a.count})`, value: a.author })),
      { name: '← indietro', value: BACK }
    ];

    const scelta = await ui.selectOpt({ choices: voci, defaultValue: voci[0].value });
    if (scelta === null || scelta === BACK) return;
    const autore = scelta;

    while (true) {
      await app.pump();
      ui.screen(`Libreria · ${autore}`, app);
      const video = app.lib.byAuthor(autore);
      // M88 — qui l'autore è già nell'intestazione: le righe non lo ripetono.
      const vociVideo = [
        ...video.map((v) => ({ name: ui.videoLine(v, { autore: false }), value: v.id })),
        { name: '← indietro', value: BACK }
      ];

      const sceltaVideo = await ui.selectOpt({ choices: vociVideo, defaultValue: vociVideo[0].value });
      if (sceltaVideo === null || sceltaVideo === BACK) break;
      await actions(app, sceltaVideo);
    }
  }
}

/**
 * "Da scaricare" mette insieme due cose che per l'utente sono la stessa: i video
 * risolti che aspettano il disco, e i link appena incollati che non hanno ancora
 * un id.
 */
async function pendingView(app) {
  while (true) {
    await app.pump();
    ui.screen('Libreria · da scaricare', app);

    // Chi sta scaricando adesso va in cima e lo dice: sta succedendo, non è in
    // attesa come gli altri. L'ordinamento è stabile, quindi dentro i due gruppi
    // resta l'ordine di `list`.
    const video = app.lib
      .list(FILTER.PENDING)
      .map((v, i) => ({ v, i }))
      .sort((a, b) => {
        const aDown = a.v.state === STATE.DOWNLOADING ? 0 : 1;
        const bDown = b.v.state === STATE.DOWNLOADING ? 0 : 1;
        return aDown - bDown || a.i - b.i;
      })
      .map((x) => x.v);

    const links = app.lib.queue().filter((q) => !q.error).map((q) => q.url);

    const voci = [];
    const scaricaTutti = video.length > 0 || links.length > 0;
    if (scaricaTutti) {
      voci.push({ name: `Scarica tutti (${video.length + links.length})`, value: 'all' });
    } else {
      console.log('Niente in attesa.\n');
    }
    for (const v of video) {
      const tag = v.state === STATE.DOWNLOADING ? 'In download' : 'In attesa';
      voci.push({ name: ui.videoLineTagged(v, tag), value: { video: v.id } });
    }
    for (const url of links) {
      voci.push({
        name: `${ui.style.yellow('○')} ${url}  ${ui.style.cyan('da risolvere')}`,
        value: { link: url }
      });
    }
    voci.push({ name: '← indietro', value: BACK });

    const scelta = await ui.selectOpt({ choices: voci, defaultValue: voci[0].value });
    if (scelta === null || scelta === BACK) return;

    if (scelta === 'all') {
      // La risoluzione si chiede **una volta** per tutto il gruppo: chiederla
      // per ognuno sarebbe un interrogatorio.
      const cap = await ui.resolveQuality(app.lib.config().quality, 'Che risoluzione, per tutti?');
      if (cap === null) continue;

      let n = 0;
      let inCorso = 0;
      for (const v of video) {
        // Chi sta già scaricando si lascia in pace: rimetterlo in coda vorrebbe
        // dire un secondo download sullo stesso video.
        if (v.state === STATE.DOWNLOADING) {
          inCorso += 1;
          continue;
        }
        const url = await app.lib.retry(v.id);
        if (url) {
          app.dl.pushWith(url, cap.maxHeight);
          n += 1;
        }
      }
      for (const url of links) {
        app.dl.pushWith(url, cap.maxHeight);
        n += 1;
      }
      const salta = inCorso > 0 ? ` (${inCorso} già in corso, lasciati stare)` : '';
      ui.ok(app, `${n} accodati${salta} — guardali in «Download rapido»`);
    } else if (scelta.video) {
      await actions(app, scelta.video);
    } else if (scelta.link) {
      await linkActions(app, scelta.link);
    }
  }
}

/**
 * I falliti, con il loro motivo: video che hanno provato e link che non si sono
 * nemmeno lasciati risolvere.
 */
async function failedView(app) {
  while (true) {
    await app.pump();
    ui.screen('Libreria · falliti', app);

    const video = app.lib.list(FILTER.FAILED);
    const linkFalliti = app.lib.queue().filter((q) => q.error);

    // Una voce di menu resta su **una** riga: l'elenco misura le voci per
    // disegnare la selezione, e una voce a due righe gli fa perdere il conto.
    const voci = [
      ...video.map((v) => ({
        name: `${ui.videoLine(v)}  ${ui.style.dim(ui.style.red(motivo(v.error)))}`,
        value: { video: v.id }
      })),
      ...linkFalliti.map((q) => ({
        name: `${ui.style.red('✗')} ${q.url}  ${ui.style.dim(ui.style.red(motivo(q.error?.message)))}`,
        value: { link: q.url }
      }))
    ];
    if (voci.length === 0) console.log('Nessun fallimento.\n');
    voci.push({ name: '← indietro', value: BACK });

    const scelta = await ui.selectOpt({ choices: voci, defaultValue: voci[0].value });
    if (scelta === null || scelta === BACK) return;
    if (scelta.video) await actions(app, scelta.video);
    else if (scelta.link) await linkActions(app, scelta.link);
  }
}

/**
 * Il motivo di un fallimento ridotto a una riga: quello che serve per capire
 * *quale* fallimento è. Per intero si vede fra le statistiche del video.
 */
function motivo(error) {
  let testo = String(error ?? 'motivo non registrato').replace(/[\n\r]/g, ' ');
  // Il pezzo utile di un errore di yt-dlp sta dopo `ERROR:`.
  const i = testo.indexOf('ERROR:');
  if (i !== -1) testo = testo.slice(i + 'ERROR:'.length).trim();
  const chars = [...testo];
  return chars.length > 60 ? `${chars.slice(0, 59).join('')}…` : testo;
}

/** Cosa si può fare con un link che non è (ancora) un video. */
async function linkActions(app, url) {
  ui.screen('Link in coda', app);
  console.log(`${url}\n`);
  const voci = [
    { name: 'Riprova adesso', value: 'retry' },
    { name: 'Toglilo dalla coda', value: 'dequeue' },
    { name: '← indietro', value: BACK }
  ];
  const scelta = await ui.selectOpt({ choices: voci, defaultValue: 'retry', pageSize: voci.length });
  if (scelta === 'retry') {
    const cap = await ui.resolveQuality(app.lib.config().quality, 'Che risoluzione?');
    if (cap === null) return;
    app.dl.pushWith(url, cap.maxHeight);
    ui.ok(app, 'accodato');
  } else if (scelta === 'dequeue') {
    await app.lib.dequeue(url);
    ui.ok(app, 'togliato dalla coda');
  }
}

/**
 * Le azioni su un video, contestuali al suo stato. Usata anche dalla ricerca: un
 * solo posto dove si decide cosa si può fare con un video.
 *
 * Le statistiche stanno **in alto**, sopra le voci: è quello che si vuole sapere
 * prima di decidere, e rende inutile una voce «Dettagli».
 */
export async function actions(app, id) {
  while (true) {
    await app.pump();
    const v = app.lib.get(id);
    if (!v) return;

    ui.screen(`${v.author} — ${v.title}`, app);
    statistiche(app, v);

    // Le voci si costruiscono in base allo stato: niente azioni impossibili da
    // scegliere e poi rifiutare. Senza icone, come il resto dei menu.
    const voci = [];
    if (v.state === STATE.DOWNLOADED) {
      voci.push({ name: 'Riproduci', value: 'play' });
      voci.push({ name: 'Solo audio', value: 'audio' });
    }
    if (v.state === STATE.PENDING || v.state === STATE.FAILED) {
      voci.push({ name: 'Scarica ora', value: 'download' });
    }
    voci.push({ name: v.favorite ? 'Togli dai preferiti' : 'Aggiungi ai preferiti', value: 'favorite' });
    voci.push({ name: v.archived ? 'Ripristina' : 'Archivia', value: 'archive' });
    voci.push({ name: 'Togli dalla libreria', value: 'remove' });
    voci.push({ name: '← indietro', value: BACK });

    const scelta = await ui.selectOpt({ choices: voci, defaultValue: voci[0].value, pageSize: voci.length });
    if (scelta === null || scelta === BACK) return;

    try {
      if (scelta === 'play') {
        await app.lib.play(id, core.PLAYBACK_MODE.VIDEO);
        ui.ok(app, 'VLC avviato');
      } else if (scelta === 'audio') {
        await app.lib.play(id, core.PLAYBACK_MODE.AUDIO);
        ui.ok(app, 'VLC avviato (solo audio)');
      } else if (scelta === 'download') {
        // Se la qualità è «chiedi ogni volta», si chiede adesso: è il momento in
        // cui il download parte.
        const cap = await ui.resolveQuality(app.lib.config().quality, 'Che risoluzione?');
        if (cap === null) continue;
        const url = await app.lib.retry(id);
        if (url) {
          app.dl.pushWith(url, cap.maxHeight);
          ui.ok(app, 'accodato — guardalo in «Download rapido»');
        } else {
          ui.err(app, 'questo video non ha un URL da riscaricare');
        }
      } else if (scelta === 'favorite') {
        const on = !v.favorite;
        await app.lib.setFavorite(id, on);
        ui.ok(app, on ? 'nei preferiti' : 'togliato dai preferiti');
      } else if (scelta === 'archive') {
        const on = !v.archived;
        await app.lib.setArchived(id, on);
        ui.ok(app, on ? 'archiviato' : 'ripristinato');
      } else if (scelta === 'remove') {
        const conferma = await confirmOpt(`Togliere «${v.title}» dalla libreria?`, false);
        if (conferma) {
          const ancheFile =
            v.state === STATE.DOWNLOADED && (await confirmOpt('Cancellare anche il file dal disco?', false));
          await app.lib.remove(id, ancheFile);
          ui.ok(app, ancheFile ? 'togliato, file cancellato' : 'togliato dalla libreria');
          return;
        }
      }
    } catch (e) {
      if (ui.isInterruzione(e)) throw e; // Ctrl-C attraversa: esce dal programma (M89)
      ui.err(app, e);
    }
  }
}

/** Le statistiche del video, sopra le voci del menu. */
function statistiche(app, v) {
  const riga = (k, val) => console.log(`  ${ui.style.dim(String(k).padEnd(12))} ${val}`);
  riga('stato', ui.style.cyan(core.stateLabel(v.state)));
  riga('durata', core.durationLabel(v.durationSeconds));
  riga('pubblicato', v.uploadDate ?? '—');
  riga(
    'risoluzione',
    v.width && v.height
      ? `${v.width}×${v.height}${v.fps ? ` @ ${Math.round(v.fps)}fps` : ''}`
      : '—'
  );
  if (v.sizeBytes > 0) riga('dimensione', ui.sizeLabel(v.sizeBytes));
  riga('url', v.url);
  if (v.state === STATE.DOWNLOADED) {
    const file = app.lib.filePath(v);
    const manca = existsSync(file) ? '' : ui.style.red("  (non è al suo posto)");
    riga('file', `${file}${manca}`);
  }
  if (v.tags.length > 0) riga('tag', v.tags.slice(0, 8).join(', '));
  if (v.error) {
    riga('errore', ui.style.red(v.error));
    riga('tentativi', String(v.attempts));
  }
  console.log();
}
